File: aws/dedicated-network-connectivity/code/cdk-python/hybrid_connectivity_stack.py
"""
Hybrid Cloud Connectivity Stack

This stack implements a complete hybrid cloud connectivity solution using:
- Multiple VPCs (Production, Development, Shared Services)
- Transit Gateway with VPC attachments
- Direct Connect Gateway
- Route 53 Resolver endpoints
- CloudWatch monitoring and alerting
- Security controls and VPC Flow Logs

Author: Recipe Generator
Version: 1.0
"""

from typing import Dict, List, Optional
from constructs import Construct
import aws_cdk as cdk
from aws_cdk import (
    Stack,
    StackProps,
    CfnOutput,
    Duration,
    aws_ec2 as ec2,
    aws_directconnect as directconnect,
    aws_route53resolver as route53resolver,
    aws_logs as logs,
    aws_iam as iam,
    aws_cloudwatch as cloudwatch,
    aws_sns as sns,
    aws_sns_subscriptions as sns_subscriptions,
)


class HybridConnectivityStack(Stack):
    """
    AWS CDK Stack for Hybrid Cloud Connectivity with Direct Connect
    
    This stack creates a comprehensive hybrid cloud connectivity solution
    with multiple VPCs, Transit Gateway, Direct Connect Gateway, and
    supporting infrastructure for DNS resolution and monitoring.
    """
    
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        project_id: str,
        environment_name: str,
        on_premises_asn: int,
        aws_asn: int,
        on_premises_cidr: str,
        **kwargs
    ) -> None:
        """
        Initialize the Hybrid Connectivity Stack.
        
        Args:
            scope: The parent construct
            construct_id: Unique identifier for this stack
            project_id: Project identifier for resource naming
            environment_name: Environment name (production, staging, etc.)
            on_premises_asn: BGP ASN for on-premises network
            aws_asn: BGP ASN for AWS side
            on_premises_cidr: CIDR block for on-premises network
            **kwargs: Additional stack properties
        """
        super().__init__(scope, construct_id, **kwargs)
        
        self.project_id = project_id
        self.environment_name = environment_name
        self.on_premises_asn = on_premises_asn
        self.aws_asn = aws_asn
        self.on_premises_cidr = on_premises_cidr
        
        # Create VPC infrastructure
        self.vpcs = self._create_vpcs()
        
        # Create Transit Gateway
        self.transit_gateway = self._create_transit_gateway()
        
        # Create VPC attachments to Transit Gateway
        self.vpc_attachments = self._create_vpc_attachments()
        
        # Create Direct Connect Gateway
        self.dx_gateway = self._create_direct_connect_gateway()
        
        # Create Transit Gateway Direct Connect attachment
        self.dx_attachment = self._create_dx_attachment()
        
        # Set up routing
        self._configure_routing()
        
        # Create Route 53 Resolver endpoints
        self.resolver_endpoints = self._create_resolver_endpoints()
        
        # Create VPC Flow Logs
        self._create_vpc_flow_logs()
        
        # Create CloudWatch monitoring
        self._create_monitoring()
        
        # Create stack outputs
        self._create_outputs()
    
    def _create_vpcs(self) -> Dict[str, ec2.Vpc]:
        """
        Create multiple VPCs for the hybrid connectivity solution.
        
        Returns:
            Dictionary of VPC names to VPC constructs
        """
        vpcs = {}
        
        # VPC configurations
        vpc_configs = {
            'production': {
                'cidr': '10.1.0.0/16',
                'description': 'Production VPC for hybrid connectivity'
            },
            'development': {
                'cidr': '10.2.0.0/16',
                'description': 'Development VPC for hybrid connectivity'
            },
            'shared-services': {
                'cidr': '10.3.0.0/16',
                'description': 'Shared Services VPC for hybrid connectivity'
            }
        }
        
        for vpc_name, config in vpc_configs.items():
            vpc = ec2.Vpc(
                self,
                f"{vpc_name.title().replace('-', '')}Vpc",
                vpc_name=f"{vpc_name}-vpc-{self.project_id}",
                ip_addresses=ec2.IpAddresses.cidr(config['cidr']),
                max_azs=2,
                subnet_configuration=[
                    ec2.SubnetConfiguration(
                        subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS,
                        name=f"{vpc_name}-private",
                        cidr_mask=24
                    ),
                    ec2.SubnetConfiguration(
                        subnet_type=ec2.SubnetType.PUBLIC,
                        name=f"{vpc_name}-public",
                        cidr_mask=24
                    )
                ],
                enable_dns_hostnames=True,
                enable_dns_support=True,
            )
            
            # Add tags
            cdk.Tags.of(vpc).add("Name", f"{vpc_name.title()}-VPC-{self.project_id}")
            cdk.Tags.of(vpc).add("VpcType", vpc_name)
            cdk.Tags.of(vpc).add("Description", config['description'])
            
            vpcs[vpc_name] = vpc
        
        return vpcs
    
    def _create_transit_gateway(self) -> ec2.CfnTransitGateway:
        """
        Create Transit Gateway for centralized routing.
        
        Returns:
            Transit Gateway construct
        """
        transit_gateway = ec2.CfnTransitGateway(
            self,
            "TransitGateway",
            amazon_side_asn=self.aws_asn,
            description=f"Corporate hybrid connectivity gateway - {self.project_id}",
            default_route_table_association="enable",
            default_route_table_propagation="enable",
            dns_support="enable",
            vpn_ecmp_support="enable",
            multicast_support="disable",
            tags=[
                cdk.CfnTag(key="Name", value=f"corporate-tgw-{self.project_id}")
            ]
        )
        
        return transit_gateway
    
    def _create_vpc_attachments(self) -> Dict[str, ec2.CfnTransitGatewayVpcAttachment]:
        """
        Create VPC attachments to Transit Gateway.
        
        Returns:
            Dictionary of VPC attachment constructs
        """
        attachments = {}
        
        for vpc_name, vpc in self.vpcs.items():
            # Get private subnets for attachment
            private_subnets = [
                subnet for subnet in vpc.private_subnets[:1]  # Use first private subnet
            ]
            
            attachment = ec2.CfnTransitGatewayVpcAttachment(
                self,
                f"{vpc_name.title().replace('-', '')}Attachment",
                transit_gateway_id=self.transit_gateway.ref,
                vpc_id=vpc.vpc_id,
                subnet_ids=[subnet.subnet_id for subnet in private_subnets],
                tags=[
                    cdk.CfnTag(key="Name", value=f"{vpc_name.title()}-TGW-Attachment")
                ]
            )
            
            attachments[vpc_name] = attachment
        
        return attachments
    
    def _create_direct_connect_gateway(self) -> directconnect.CfnDirectConnectGateway:
        """
        Create Direct Connect Gateway.
        
        Returns:
            Direct Connect Gateway construct
        """
        dx_gateway = directconnect.CfnDirectConnectGateway(
            self,
            "DirectConnectGateway",
            name=f"corporate-dx-gateway-{self.project_id}",
            amazon_side_asn=self.aws_asn
        )
        
        return dx_gateway
    
    def _create_dx_attachment(self) -> ec2.CfnTransitGatewayDirectConnectGatewayAttachment:
        """
        Create Transit Gateway Direct Connect Gateway attachment.
        
        Returns:
            Transit Gateway Direct Connect Gateway attachment
        """
        dx_attachment = ec2.CfnTransitGatewayDirectConnectGatewayAttachment(
            self,
            "DxTgwAttachment",
            transit_gateway_id=self.transit_gateway.ref,
            direct_connect_gateway_id=self.dx_gateway.ref,
            tags=[
                cdk.CfnTag(key="Name", value=f"DX-TGW-Attachment-{self.project_id}")
            ]
        )
        
        return dx_attachment
    
    def _configure_routing(self) -> None:
        """
        Configure routing between VPCs and on-premises networks.
        """
        # Add routes to on-premises networks in each VPC
        for vpc_name, vpc in self.vpcs.items():
            # Add route to on-premises CIDR via Transit Gateway
            for subnet in vpc.private_subnets:
                ec2.CfnRoute(
                    self,
                    f"{vpc_name.title().replace('-', '')}OnPremRoute{subnet.node.id}",
                    route_table_id=subnet.route_table.route_table_id,
                    destination_cidr_block=self.on_premises_cidr,
                    transit_gateway_id=self.transit_gateway.ref
                )
    
    def _create_resolver_endpoints(self) -> Dict[str, route53resolver.CfnResolverEndpoint]:
        """
        Create Route 53 Resolver endpoints for DNS resolution.
        
        Returns:
            Dictionary of resolver endpoint constructs
        """
        endpoints = {}
        
        # Create security group for resolver endpoints
        resolver_sg = ec2.SecurityGroup(
            self,
            "ResolverSecurityGroup",
            vpc=self.vpcs['shared-services'],
            description="Security group for Route 53 Resolver endpoints",
            allow_all_outbound=True
        )
        
        # Allow DNS traffic from on-premises
        resolver_sg.add_ingress_rule(
            peer=ec2.Peer.ipv4(self.on_premises_cidr),
            connection=ec2.Port.udp(53),
            description="Allow DNS UDP from on-premises"
        )
        
        resolver_sg.add_ingress_rule(
            peer=ec2.Peer.ipv4(self.on_premises_cidr),
            connection=ec2.Port.tcp(53),
            description="Allow DNS TCP from on-premises"
        )
        
        # Allow DNS traffic from VPCs
        for vpc_name, vpc in self.vpcs.items():
            resolver_sg.add_ingress_rule(
                peer=ec2.Peer.ipv4(vpc.vpc_cidr_block),
                connection=ec2.Port.udp(53),
                description=f"Allow DNS UDP from {vpc_name} VPC"
            )
        
        # Create inbound resolver endpoint
        shared_vpc = self.vpcs['shared-services']
        private_subnet = shared_vpc.private_subnets[0]
        
        inbound_endpoint = route53resolver.CfnResolverEndpoint(
            self,
            "InboundResolverEndpoint",
            direction="INBOUND",
            ip_addresses=[
                route53resolver.CfnResolverEndpoint.IpAddressRequestProperty(
                    subnet_id=private_subnet.subnet_id,
                    ip="10.3.1.100"
                )
            ],
            security_group_ids=[resolver_sg.security_group_id],
            name=f"Inbound-Resolver-{self.project_id}"
        )
        
        # Create outbound resolver endpoint
        outbound_endpoint = route53resolver.CfnResolverEndpoint(
            self,
            "OutboundResolverEndpoint",
            direction="OUTBOUND",
            ip_addresses=[
                route53resolver.CfnResolverEndpoint.IpAddressRequestProperty(
                    subnet_id=private_subnet.subnet_id,
                    ip="10.3.1.101"
                )
            ],
            security_group_ids=[resolver_sg.security_group_id],
            name=f"Outbound-Resolver-{self.project_id}"
        )
        
        endpoints['inbound'] = inbound_endpoint
        endpoints['outbound'] = outbound_endpoint
        
        return endpoints
    
    def _create_vpc_flow_logs(self) -> None:
        """
        Create VPC Flow Logs for all VPCs.
        """
        # Create CloudWatch log group for VPC Flow Logs
        flow_log_group = logs.LogGroup(
            self,
            "VpcFlowLogsGroup",
            log_group_name=f"/aws/vpc/flowlogs-{self.project_id}",
            retention=logs.RetentionDays.ONE_MONTH,
            removal_policy=cdk.RemovalPolicy.DESTROY
        )
        
        # Create IAM role for VPC Flow Logs
        flow_logs_role = iam.Role(
            self,
            "VpcFlowLogsRole",
            assumed_by=iam.ServicePrincipal("vpc-flow-logs.amazonaws.com"),
            inline_policies={
                "FlowLogsDeliveryRolePolicy": iam.PolicyDocument(
                    statements=[
                        iam.PolicyStatement(
                            effect=iam.Effect.ALLOW,
                            actions=[
                                "logs:CreateLogGroup",
                                "logs:CreateLogStream",
                                "logs:PutLogEvents",
                                "logs:DescribeLogGroups",
                                "logs:DescribeLogStreams"
                            ],
                            resources=["*"]
                        )
                    ]
                )
            }
        )
        
        # Create VPC Flow Logs for each VPC
        for vpc_name, vpc in self.vpcs.items():
            ec2.CfnFlowLog(
                self,
                f"{vpc_name.title().replace('-', '')}FlowLog",
                resource_type="VPC",
                resource_id=vpc.vpc_id,
                traffic_type="ALL",
                log_destination_type="cloud-watch-logs",
                log_group_name=flow_log_group.log_group_name,
                deliver_logs_permission_arn=flow_logs_role.role_arn
            )
    
    def _create_monitoring(self) -> None:
        """
        Create CloudWatch monitoring and alerting.
        """
        # Create SNS topic for alerts
        alert_topic = sns.Topic(
            self,
            "DirectConnectAlerts",
            display_name=f"Direct Connect Alerts - {self.project_id}",
            topic_name=f"dx-alerts-{self.project_id}"
        )
        
        # Create CloudWatch Dashboard
        dashboard = cloudwatch.Dashboard(
            self,
            "DirectConnectDashboard",
            dashboard_name=f"DirectConnect-{self.project_id}",
            period_override=cloudwatch.PeriodOverride.INHERIT,
            start="P1D",
            widgets=[
                [
                    cloudwatch.GraphWidget(
                        title="Direct Connect Connection State",
                        left=[
                            cloudwatch.Metric(
                                namespace="AWS/DX",
                                metric_name="ConnectionState",
                                dimensions_map={
                                    "ConnectionId": "CONNECTION_ID_PLACEHOLDER"
                                },
                                statistic="Maximum",
                                period=Duration.minutes(5)
                            )
                        ],
                        width=12,
                        height=6
                    )
                ],
                [
                    cloudwatch.GraphWidget(
                        title="Direct Connect Bandwidth",
                        left=[
                            cloudwatch.Metric(
                                namespace="AWS/DX",
                                metric_name="ConnectionBpsEgress",
                                dimensions_map={
                                    "ConnectionId": "CONNECTION_ID_PLACEHOLDER"
                                },
                                statistic="Average",
                                period=Duration.minutes(5)
                            )
                        ],
                        right=[
                            cloudwatch.Metric(
                                namespace="AWS/DX",
                                metric_name="ConnectionBpsIngress",
                                dimensions_map={
                                    "ConnectionId": "CONNECTION_ID_PLACEHOLDER"
                                },
                                statistic="Average",
                                period=Duration.minutes(5)
                            )
                        ],
                        width=12,
                        height=6
                    )
                ]
            ]
        )
        
        # Create CloudWatch Alarm for connection state
        # Note: This alarm would need to be configured with actual connection ID
        # after the physical Direct Connect connection is established
        
    def _create_outputs(self) -> None:
        """
        Create CloudFormation outputs for key resources.
        """
        # VPC outputs
        for vpc_name, vpc in self.vpcs.items():
            CfnOutput(
                self,
                f"{vpc_name.title().replace('-', '')}VpcId",
                value=vpc.vpc_id,
                description=f"ID of the {vpc_name} VPC",
                export_name=f"{self.stack_name}-{vpc_name.replace('-', '')}-vpc-id"
            )
        
        # Transit Gateway outputs
        CfnOutput(
            self,
            "TransitGatewayId",
            value=self.transit_gateway.ref,
            description="ID of the Transit Gateway",
            export_name=f"{self.stack_name}-transit-gateway-id"
        )
        
        # Direct Connect Gateway outputs
        CfnOutput(
            self,
            "DirectConnectGatewayId",
            value=self.dx_gateway.ref,
            description="ID of the Direct Connect Gateway",
            export_name=f"{self.stack_name}-dx-gateway-id"
        )
        
        # Resolver endpoints outputs
        CfnOutput(
            self,
            "InboundResolverEndpointId",
            value=self.resolver_endpoints['inbound'].ref,
            description="ID of the inbound Route 53 Resolver endpoint",
            export_name=f"{self.stack_name}-inbound-resolver-id"
        )
        
        CfnOutput(
            self,
            "OutboundResolverEndpointId",
            value=self.resolver_endpoints['outbound'].ref,
            description="ID of the outbound Route 53 Resolver endpoint",
            export_name=f"{self.stack_name}-outbound-resolver-id"
        )
        
        # Configuration outputs
        CfnOutput(
            self,
            "OnPremisesCidr",
            value=self.on_premises_cidr,
            description="CIDR block for on-premises network"
        )
        
        CfnOutput(
            self,
            "OnPremisesAsn",
            value=str(self.on_premises_asn),
            description="BGP ASN for on-premises network"
        )
        
        CfnOutput(
            self,
            "AwsAsn",
            value=str(self.aws_asn),
            description="BGP ASN for AWS side"
        )
        
        # DNS server IP addresses
        CfnOutput(
            self,
            "InboundDnsServerIp",
            value="10.3.1.100",
            description="IP address of inbound DNS resolver endpoint"
        )
        
        CfnOutput(
            self,
            "OutboundDnsServerIp",
            value="10.3.1.101",
            description="IP address of outbound DNS resolver endpoint"
        )
        
        # BGP configuration template
        CfnOutput(
            self,
            "BgpConfigurationNote",
            value="See README.md for BGP configuration template and virtual interface setup instructions",
            description="BGP configuration instructions"
        )

File: aws/dedicated-network-connectivity/code/cdk-typescript/lib/hybrid-connectivity-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as dx from 'aws-cdk-lib/aws-directconnect';
import * as route53resolver from 'aws-cdk-lib/aws-route53resolver';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

export interface HybridConnectivityStackProps extends cdk.StackProps {
  onPremisesCidr: string;
  onPremisesAsn: number;
  awsAsn: number;
  privateVifVlan: number;
  transitVifVlan: number;
  projectId: string;
  enableDnsResolution: boolean;
  enableMonitoring: boolean;
  enableFlowLogs: boolean;
}

export class HybridConnectivityStack extends cdk.Stack {
  public readonly transitGateway: ec2.CfnTransitGateway;
  public readonly directConnectGateway: dx.CfnDirectConnectGateway;
  public readonly productionVpc: ec2.Vpc;
  public readonly developmentVpc: ec2.Vpc;
  public readonly sharedServicesVpc: ec2.Vpc;

  constructor(scope: Construct, id: string, props: HybridConnectivityStackProps) {
    super(scope, id, props);

    // Create VPC infrastructure
    this.createVpcInfrastructure(props);

    // Create Transit Gateway
    this.createTransitGateway(props);

    // Create Direct Connect Gateway
    this.createDirectConnectGateway(props);

    // Attach VPCs to Transit Gateway
    this.attachVpcsToTransitGateway(props);

    // Associate Direct Connect Gateway with Transit Gateway
    this.associateDirectConnectGatewayWithTransitGateway(props);

    // Configure DNS resolution if enabled
    if (props.enableDnsResolution) {
      this.configureDnsResolution(props);
    }

    // Configure monitoring if enabled
    if (props.enableMonitoring) {
      this.configureMonitoring(props);
    }

    // Configure VPC Flow Logs if enabled
    if (props.enableFlowLogs) {
      this.configureFlowLogs(props);
    }

    // Create outputs
    this.createOutputs(props);
  }

  private createVpcInfrastructure(props: HybridConnectivityStackProps): void {
    // Production VPC
    this.productionVpc = new ec2.Vpc(this, 'ProductionVpc', {
      cidr: '10.1.0.0/16',
      maxAzs: 3,
      natGateways: 1,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        },
        {
          cidrMask: 28,
          name: 'tgw',
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
    });

    // Development VPC
    this.developmentVpc = new ec2.Vpc(this, 'DevelopmentVpc', {
      cidr: '10.2.0.0/16',
      maxAzs: 3,
      natGateways: 1,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        },
        {
          cidrMask: 28,
          name: 'tgw',
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
    });

    // Shared Services VPC
    this.sharedServicesVpc = new ec2.Vpc(this, 'SharedServicesVpc', {
      cidr: '10.3.0.0/16',
      maxAzs: 3,
      natGateways: 1,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        },
        {
          cidrMask: 28,
          name: 'tgw',
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
    });

    // Add tags to VPCs
    cdk.Tags.of(this.productionVpc).add('Name', `Production-VPC-${props.projectId}`);
    cdk.Tags.of(this.productionVpc).add('Environment', 'Production');
    cdk.Tags.of(this.developmentVpc).add('Name', `Development-VPC-${props.projectId}`);
    cdk.Tags.of(this.developmentVpc).add('Environment', 'Development');
    cdk.Tags.of(this.sharedServicesVpc).add('Name', `Shared-Services-VPC-${props.projectId}`);
    cdk.Tags.of(this.sharedServicesVpc).add('Environment', 'Shared');
  }

  private createTransitGateway(props: HybridConnectivityStackProps): void {
    this.transitGateway = new ec2.CfnTransitGateway(this, 'TransitGateway', {
      amazonSideAsn: props.awsAsn,
      description: 'Corporate hybrid connectivity gateway',
      defaultRouteTableAssociation: 'enable',
      defaultRouteTablePropagation: 'enable',
      dnsSupport: 'enable',
      vpnEcmpSupport: 'enable',
      tags: [
        {
          key: 'Name',
          value: `corporate-tgw-${props.projectId}`,
        },
      ],
    });
  }

  private createDirectConnectGateway(props: HybridConnectivityStackProps): void {
    this.directConnectGateway = new dx.CfnDirectConnectGateway(this, 'DirectConnectGateway', {
      name: `corporate-dx-gateway-${props.projectId}`,
      amazonSideAsn: props.awsAsn,
    });
  }

  private attachVpcsToTransitGateway(props: HybridConnectivityStackProps): void {
    // Production VPC attachment
    new ec2.CfnTransitGatewayVpcAttachment(this, 'ProductionVpcAttachment', {
      transitGatewayId: this.transitGateway.ref,
      vpcId: this.productionVpc.vpcId,
      subnetIds: this.productionVpc.isolatedSubnets.map(subnet => subnet.subnetId),
      tags: [
        {
          key: 'Name',
          value: 'Prod-TGW-Attachment',
        },
      ],
    });

    // Development VPC attachment
    new ec2.CfnTransitGatewayVpcAttachment(this, 'DevelopmentVpcAttachment', {
      transitGatewayId: this.transitGateway.ref,
      vpcId: this.developmentVpc.vpcId,
      subnetIds: this.developmentVpc.isolatedSubnets.map(subnet => subnet.subnetId),
      tags: [
        {
          key: 'Name',
          value: 'Dev-TGW-Attachment',
        },
      ],
    });

    // Shared Services VPC attachment
    new ec2.CfnTransitGatewayVpcAttachment(this, 'SharedServicesVpcAttachment', {
      transitGatewayId: this.transitGateway.ref,
      vpcId: this.sharedServicesVpc.vpcId,
      subnetIds: this.sharedServicesVpc.isolatedSubnets.map(subnet => subnet.subnetId),
      tags: [
        {
          key: 'Name',
          value: 'Shared-TGW-Attachment',
        },
      ],
    });

    // Update route tables in each VPC to route on-premises traffic through Transit Gateway
    this.updateVpcRouteTables(props);
  }

  private updateVpcRouteTables(props: HybridConnectivityStackProps): void {
    const vpcs = [this.productionVpc, this.developmentVpc, this.sharedServicesVpc];
    
    vpcs.forEach((vpc, index) => {
      // Add route for on-premises traffic to private subnets
      vpc.privateSubnets.forEach((subnet, subnetIndex) => {
        new ec2.CfnRoute(this, `OnPremRoute-${index}-${subnetIndex}`, {
          routeTableId: subnet.routeTable.routeTableId,
          destinationCidrBlock: props.onPremisesCidr,
          transitGatewayId: this.transitGateway.ref,
        });
      });

      // Add route for on-premises traffic to isolated subnets
      vpc.isolatedSubnets.forEach((subnet, subnetIndex) => {
        new ec2.CfnRoute(this, `OnPremRouteIsolated-${index}-${subnetIndex}`, {
          routeTableId: subnet.routeTable.routeTableId,
          destinationCidrBlock: props.onPremisesCidr,
          transitGatewayId: this.transitGateway.ref,
        });
      });
    });
  }

  private associateDirectConnectGatewayWithTransitGateway(props: HybridConnectivityStackProps): void {
    new ec2.CfnTransitGatewayDirectConnectGatewayAttachment(this, 'DxGatewayAttachment', {
      transitGatewayId: this.transitGateway.ref,
      directConnectGatewayId: this.directConnectGateway.ref,
      tags: [
        {
          key: 'Name',
          value: `DX-TGW-Attachment-${props.projectId}`,
        },
      ],
    });
  }

  private configureDnsResolution(props: HybridConnectivityStackProps): void {
    // Create security group for resolver endpoints
    const resolverSecurityGroup = new ec2.SecurityGroup(this, 'ResolverSecurityGroup', {
      vpc: this.sharedServicesVpc,
      description: 'Security group for Route 53 Resolver endpoints',
      securityGroupName: `resolver-endpoints-sg-${props.projectId}`,
    });

    // Allow DNS traffic from on-premises
    resolverSecurityGroup.addIngressRule(
      ec2.Peer.ipv4(props.onPremisesCidr),
      ec2.Port.udp(53),
      'Allow DNS UDP from on-premises'
    );

    resolverSecurityGroup.addIngressRule(
      ec2.Peer.ipv4(props.onPremisesCidr),
      ec2.Port.tcp(53),
      'Allow DNS TCP from on-premises'
    );

    // Allow DNS traffic between VPCs
    const vpcCidrs = ['10.1.0.0/16', '10.2.0.0/16', '10.3.0.0/16'];
    vpcCidrs.forEach(cidr => {
      resolverSecurityGroup.addIngressRule(
        ec2.Peer.ipv4(cidr),
        ec2.Port.udp(53),
        `Allow DNS UDP from ${cidr}`
      );
      resolverSecurityGroup.addIngressRule(
        ec2.Peer.ipv4(cidr),
        ec2.Port.tcp(53),
        `Allow DNS TCP from ${cidr}`
      );
    });

    // Create inbound resolver endpoint
    new route53resolver.CfnResolverEndpoint(this, 'InboundResolverEndpoint', {
      direction: 'INBOUND',
      ipAddresses: [
        {
          subnetId: this.sharedServicesVpc.isolatedSubnets[0].subnetId,
          ip: '10.3.1.100',
        },
      ],
      securityGroupIds: [resolverSecurityGroup.securityGroupId],
      name: `Inbound-Resolver-${props.projectId}`,
    });

    // Create outbound resolver endpoint
    new route53resolver.CfnResolverEndpoint(this, 'OutboundResolverEndpoint', {
      direction: 'OUTBOUND',
      ipAddresses: [
        {
          subnetId: this.sharedServicesVpc.isolatedSubnets[0].subnetId,
          ip: '10.3.1.101',
        },
      ],
      securityGroupIds: [resolverSecurityGroup.securityGroupId],
      name: `Outbound-Resolver-${props.projectId}`,
    });
  }

  private configureMonitoring(props: HybridConnectivityStackProps): void {
    // Create CloudWatch dashboard
    const dashboard = new cloudwatch.Dashboard(this, 'DirectConnectDashboard', {
      dashboardName: `DirectConnect-${props.projectId}`,
    });

    // Note: Direct Connect metrics would be added here if connection ID is available
    // For now, we'll create Transit Gateway metrics
    const transitGatewayMetrics = new cloudwatch.GraphWidget({
      title: 'Transit Gateway Metrics',
      left: [
        new cloudwatch.Metric({
          namespace: 'AWS/TransitGateway',
          metricName: 'BytesIn',
          dimensionsMap: {
            TransitGateway: this.transitGateway.ref,
          },
          statistic: 'Sum',
        }),
        new cloudwatch.Metric({
          namespace: 'AWS/TransitGateway',
          metricName: 'BytesOut',
          dimensionsMap: {
            TransitGateway: this.transitGateway.ref,
          },
          statistic: 'Sum',
        }),
      ],
    });

    dashboard.addWidgets(transitGatewayMetrics);

    // Create CloudWatch alarm for monitoring
    new cloudwatch.Alarm(this, 'TransitGatewayAlarm', {
      alarmName: `TGW-HighTraffic-${props.projectId}`,
      alarmDescription: 'High traffic on Transit Gateway',
      metric: new cloudwatch.Metric({
        namespace: 'AWS/TransitGateway',
        metricName: 'BytesIn',
        dimensionsMap: {
          TransitGateway: this.transitGateway.ref,
        },
        statistic: 'Sum',
        period: cdk.Duration.minutes(5),
      }),
      threshold: 1000000000, // 1GB in bytes
      evaluationPeriods: 2,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
    });
  }

  private configureFlowLogs(props: HybridConnectivityStackProps): void {
    // Create IAM role for VPC Flow Logs
    const flowLogsRole = new iam.Role(this, 'FlowLogsRole', {
      assumedBy: new iam.ServicePrincipal('vpc-flow-logs.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/VPCFlowLogsDeliveryRolePolicy'),
      ],
    });

    // Create CloudWatch log group
    const logGroup = new logs.LogGroup(this, 'VpcFlowLogsGroup', {
      logGroupName: `/aws/vpc/flowlogs-${props.projectId}`,
      retention: logs.RetentionDays.ONE_MONTH,
    });

    // Create flow logs for each VPC
    const vpcs = [
      { vpc: this.productionVpc, name: 'Production' },
      { vpc: this.developmentVpc, name: 'Development' },
      { vpc: this.sharedServicesVpc, name: 'SharedServices' },
    ];

    vpcs.forEach(({ vpc, name }) => {
      new ec2.CfnFlowLog(this, `${name}FlowLog`, {
        resourceType: 'VPC',
        resourceId: vpc.vpcId,
        trafficType: 'ALL',
        logDestinationType: 'cloud-watch-logs',
        logGroupName: logGroup.logGroupName,
        deliverLogsPermissionArn: flowLogsRole.roleArn,
        tags: [
          {
            key: 'Name',
            value: `${name}-FlowLog-${props.projectId}`,
          },
        ],
      });
    });
  }

  private createOutputs(props: HybridConnectivityStackProps): void {
    // VPC outputs
    new cdk.CfnOutput(this, 'ProductionVpcId', {
      value: this.productionVpc.vpcId,
      description: 'Production VPC ID',
    });

    new cdk.CfnOutput(this, 'DevelopmentVpcId', {
      value: this.developmentVpc.vpcId,
      description: 'Development VPC ID',
    });

    new cdk.CfnOutput(this, 'SharedServicesVpcId', {
      value: this.sharedServicesVpc.vpcId,
      description: 'Shared Services VPC ID',
    });

    // Transit Gateway outputs
    new cdk.CfnOutput(this, 'TransitGatewayId', {
      value: this.transitGateway.ref,
      description: 'Transit Gateway ID',
    });

    // Direct Connect Gateway outputs
    new cdk.CfnOutput(this, 'DirectConnectGatewayId', {
      value: this.directConnectGateway.ref,
      description: 'Direct Connect Gateway ID',
    });

    // DNS resolver endpoint IPs
    if (props.enableDnsResolution) {
      new cdk.CfnOutput(this, 'InboundResolverEndpointIp', {
        value: '10.3.1.100',
        description: 'Inbound DNS Resolver Endpoint IP',
      });

      new cdk.CfnOutput(this, 'OutboundResolverEndpointIp', {
        value: '10.3.1.101',
        description: 'Outbound DNS Resolver Endpoint IP',
      });
    }

    // BGP configuration guidance
    new cdk.CfnOutput(this, 'BgpConfiguration', {
      value: `Configure BGP with ASN ${props.onPremisesAsn} on your on-premises router`,
      description: 'BGP Configuration Instructions',
    });

    // Virtual Interface configuration
    new cdk.CfnOutput(this, 'VirtualInterfaceConfig', {
      value: `Create Transit VIF with VLAN ${props.transitVifVlan} connected to DX Gateway ${this.directConnectGateway.ref}`,
      description: 'Virtual Interface Configuration',
    });
  }
}